### PR TITLE
tweet-api flask endpoint

### PR DESCRIPTION
**Pull request by danasaur**
connected front-end to pre-built pytorch twitter sentiment classifier model

### some cleanup TODOS remaining:

- [ ] I want to update the test file to use pytest and add some tests and test-payloads
- [ ] serve api template or tweet-api on a different port so there's no conflict (or remove api?)
- [ ] pin transformers, torch and scipy in requirements.txt
- [ ] for @soulshake , fix issue with needing to comment out volumes for front-end
- [ ] figure out how to set the tweet-api app to refresh / check for updates to the code so I can iterate on the code and test it without taking the app down (I had this before, I think a gunicorn param or something)

### Architecture design decisions to discuss:

**Response Payload:** 

> We could demonstrate architecture design trade-offs between having the api return the most likely class (simple and the design we have here) OR sending back the scores for each class. Example response:
> 
> `{"joy": .92, "sadness": .02, "optimism": .83, "anger": .03} `
> 
> This would force the caller to do logic on their side but gives them a lot of flexibility to change the UI without requiring an update to the api. Examples: 
> 
> - filter to the most likely class and display that
> - only accept a classification if the score is above a threshold of say .9
> - filter to any classes that have a score above a threshold, say .8
> - display a ranked list of the classes as suggestions in a drop-down menu

> Alternatives? Return both? Two separate endpoints? Force the caller to specify what they want in their request payload with a parameter like "verbose"?

**Loading the model files into memory:**

> This implementation loads pre-built model files into memory when the app starts up (upon import of inference.py). Those are baked into the docker image from the requirements.txt

> We could cover architecture design trade-offs of alternative approaches for custom-built model archive files. For example:
> - bake them into the docker image
> - download via an api call to s3/blob storage etc. when the app starts
> - volume mount them when developing locally with docker-compose and store them on a persistent volume for running on kubernetes?
> - other options?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,129 @@
 frontend/public/env.js
+
+
```
Comment (jpetazzo) [inline .gitignore:3]: Do we really need all these?
I tend to put only the necessary gitignores to avoid being bitten by "whoops my legit file wasn't added because it clashes with this 100+ lines gitignore/dockerignore/rsyncignore file"...

File: .gitignore [hunk 2]
```diff
@@ -1 +1,129 @@
...
+.webassets-cache
+
+# Scrapy stuff:
+.scrapy
+
+# Sphinx documentation
+docs/_build/
+
+# PyBuilder
+target/
```
Comment (jpetazzo) [inline .gitignore:89]: Especially that one...

File: .gitignore [hunk 3]
```diff
@@ -1 +1,129 @@
...
+
+# Jupyter Notebook
+.ipynb_checkpoints
+
+# IPython
+profile_default/
+ipython_config.py
+
+# pyenv
+.python-version
```
Comment (jpetazzo) [inline .gitignore:99]: And that one...

File: .gitignore [hunk 4]
```diff
@@ -1 +1,129 @@
...
+# celery beat schedule file
+celerybeat-schedule
+
+# SageMath parsed files
+*.sage.py
+
+# Environments
+.env
+.venv
+env/
```
Comment (jpetazzo) [inline .gitignore:110]: And that one...

File: .gitignore [hunk 5]
```diff
@@ -1 +1,129 @@
...
+
+# Spyder project settings
+.spyderproject
+.spyproject
+
+# Rope project settings
+.ropeproject
+
+# mkdocs documentation
+/site
```
Comment (jpetazzo) [inline .gitignore:124]: And that one...

File: docker-compose.yml [hunk 6]
```diff
@@ -12,6 +12,18 @@ services:
     ports:
       - 5000:5000
 
+  tweet-api:
```
Comment (jpetazzo) [inline docker-compose.yml:15]: Would it make sense to just call it `api`? Simpler and faster to type; and what we're doing isn't tied to Twitter, is it?

File: docker-compose.yml [hunk 7]
```diff
@@ -12,6 +12,18 @@ services:
...
+    build:
```
Comment (jpetazzo) [inline docker-compose.yml:16]: Why is this a build arg? 🤔
It looks like this should be a runtime variable at best.

File: docker-compose.yml [hunk 8]
```diff
@@ -12,6 +12,18 @@ services:
...
+      args:
+        FLASK_DEBUG: ${FLASK_DEBUG:-0}
+      context: tweet-api
+    environment:
+      FLASK_DEBUG: ${FLASK_DEBUG:-0}
```
Comment (jpetazzo) [inline docker-compose.yml:21]: I'd suggest to just pass the variable as-is, i.e.:
```yaml
environment:
- FLASK_DEBUG
```

File: docker-compose.yml [hunk 9]
```diff
@@ -12,6 +12,18 @@ services:
...
+    volumes:
+      - ./tweet-api/:/app
+    ports:
+      - 5000:5000
```
Comment (jpetazzo) [inline docker-compose.yml:25]: However, here I'd suggest `- ${PORT-5000}:5000` to make it easier to stage multiple versions side by side

Comment (jpetazzo) [reply docker-compose.yml:25]: Or maybe `- ${API_PORT-5000}:5000` since we also have a frontend port below 👍🏻

File: docker-compose.yml [hunk 10]
```diff
@@ -21,7 +33,7 @@ services:
     environment:
       DIST_DIR: /app/public
       API_URL: http://localhost:5000
-    volumes:
-      - ./frontend/:/app
+    # volumes:
+    #   - ./frontend/:/app
     ports:
-      - 3000:3000
+      - 3000:3000
```
Comment (jpetazzo) [inline docker-compose.yml:39]: Ditto, maybe `${FRONTEND_PORT:3000}` and reflect it in `API_URL` above - maybe?

File: tweet-api/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# syntax=docker/dockerfile:1
+
+ARG FLASK_DEBUG=0
```
Comment (jpetazzo) [inline tweet-api/Dockerfile:3]: Do we need this?  (Cf comment above about using a build arg for FLASK_DEBUG)

File: tweet-api/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+FROM python:3.9-slim-buster as development
+ENV FLASK_APP=api/app.py
+ENV FLASK_DEBUG=${FLASK_DEBUG}
```
Comment (jpetazzo) [inline tweet-api/Dockerfile:7]: Do we need this?

File: tweet-api/api/app.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+logging.getLogger("flask_cors").level = logging.DEBUG
+app = Flask(__name__)
+CORS(app)
+
+app.json_encoder = _NumpyEncoder
+
+
+@app.route("/")
+def hello_world():
+    return "<p>Hello, World!</p>"
```
Comment (jpetazzo) [inline tweet-api/api/app.py:39]: Suggestion: use the same kind of banner as I do for Dockercoins (https://github.com/jpetazzo/container.training/blob/main/dockercoins/rng/rng.py#L18) because it shows the container/pod identity

File: tweet-api/api/app.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+@app.route("/analyze", methods=["POST"])
+def analyze():
+    start = time.time()
+
+    errors = []
+    # sentiment = "empty"
+    text = request.json.get("text")
+    if text:
+        # app.logger.info(f"Analyzing: {text}")
```
Comment (jpetazzo) [inline tweet-api/api/app.py:55]: remove the commented out section?

File: tweet-api/api/app.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        # elif len(text) == 3:
+        #     sentiment = "excited"
+        #     errors.append({"message": "nice"})
+        # else:
+        #     sentiment = "happy"
+        sentiment = inference.predict(text)
+    else:
+        errors.append({"message": "no text received in request json"})
+
+    time.sleep(1)
```
Comment (jpetazzo) [inline tweet-api/api/app.py:67]: Do we still need to pad the execution time?

File: tweet-api/api/app.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    end = time.time()
+    duration = round((end - start) * 1000, 4)
+    # ret = {"data": {"sentiment": sentiment, "duration": duration, "errors": errors}}
```
Comment (jpetazzo) [inline tweet-api/api/app.py:70]: ✂️ 🙂

File: tweet-api/api/inference.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
+import csv
+import urllib.request
+
+import numpy as np
+from scipy.special import softmax
+from transformers import AutoModelForSequenceClassification, AutoTokenizer
+
+task='emotion'
```
Comment (jpetazzo) [inline tweet-api/api/inference.py:8]: Maybe run the code through `black` to get consistent formatting etc (@soulshake wdyt?)

File: tweet-api/api/inference.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+MODEL = f"cardiffnlp/twitter-roberta-base-{task}"
+tokenizer = AutoTokenizer.from_pretrained(MODEL)
+
+# download label mapping
+mapping_link = f"https://raw.githubusercontent.com/cardiffnlp/tweeteval/main/datasets/{task}/mapping.txt"
+with urllib.request.urlopen(mapping_link) as f:
+    html = f.read().decode('utf-8').split("\n")
+    csvreader = csv.reader(html, delimiter='\t')
```
Comment (jpetazzo) [inline tweet-api/api/inference.py:16]: Would it be simpler to skip the csvreader here - since it's just 4 lines with `<int>\t<label>`?
(Not super important, but it gives the impression that we need to "use the big guns" - the `csv` package - to do a very simple task!)

File: tweet-api/api/inference.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+labels = [row[1] for row in csvreader if len(row) > 1]
+
+def _preprocess(text):
+    new_text = []
+    for t in text.split(" "):
```
Comment (jpetazzo) [inline tweet-api/api/inference.py:21]: Any particular reason why we use `.split(" ")` instead of `.split()` ?

File: tweet-api/requirements.txt [hunk 20]
```diff
@@ -0,0 +1,5 @@
+Flask==2.2.2
+Flask-Cors==3.0.10
```
Comment (jpetazzo) [inline tweet-api/requirements.txt:2]: This doesn't seem to be used, or am I missing it somewhere else?

File: tweet-api/service.yaml [hunk 21]
```diff
@@ -0,0 +1,84 @@
+apiVersion: networking.k8s.io/v1
```
Comment (jpetazzo) [inline tweet-api/service.yaml:1]: I think that whole file needs to be simplified a lot because I see lots of extraneous things, but I can take care of it later 👍🏻

File: tweet-api/tests/test_analyze.py [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+import unittest
+
+import requests
+
+class TestStringMethods(unittest.TestCase):
+
+    def test_analyze(self):
+        with open("./sample_requests/request.json") as f:
+            # python 3.7 dicts are ordered now, but some other libraries like pandas don't respect that
+            # so it's safest to stick with an OrderedDict since the order of the features for our input vector matter
```
Comment (jpetazzo) [inline tweet-api/tests/test_analyze.py:12]: oh why does the order matter? (in this specific case there is only one key anyway so it wouldn't matter, but I'd love to understand why in the general case!)

**Review by jpetazzo — COMMENTED** [review]
I left a bunch of comments; feel free to discuss here or this afternoon in person! :)

File: tweet-api/requirements.txt (hunk 20 above)
Comment (danasaur) [reply tweet-api/requirements.txt:2]: @soulshake

File: tweet-api/service.yaml (hunk 21 above)
Comment (danasaur) [reply tweet-api/service.yaml:1]: @soulshake

File: docker-compose.yml (hunk 6 above)
Comment (soulshake) [reply docker-compose.yml:15]: @danasaur feel free to name yours `api` and rename mine to, say, `mock-api` or `fake-api` or something

File: docker-compose.yml (hunk 7 above)
Comment (soulshake) [reply docker-compose.yml:16]: That was my bad; `FLASK_DEBUG` can be removed as a build arg

File: docker-compose.yml (hunk 8 above)
Comment (soulshake) [reply docker-compose.yml:21]: I think you get a warning if it's not set in that case. I'd leave it as-is (worth double-checking though).
(Actually, it should default to `1` and not `0`)

File: tweet-api/api/inference.py (hunk 17 above)
Comment (soulshake) [reply tweet-api/api/inference.py:8]: +1 for running Black (there should be a vsCode plugin @danasaur)

File: tweet-api/api/inference.py [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+        t = '@user' if t.startswith('@') and len(t) > 1 else t
+        t = 'http' if t.startswith('http') else t
+        new_text.append(t)
+    return " ".join(new_text)
+
+def predict(text):
```
Comment (soulshake) [inline tweet-api/api/inference.py:27]: A few comments here would go a long way for us non-data-scientists :)

File: tweet-api/requirements.txt (hunk 20 above)
Comment (soulshake) [reply tweet-api/requirements.txt:2]: It's used [here](https://github.com/soulshake/kuberpoets/pull/5/files#diff-785e186ac3ec10909cdeffef9a4f0f859060388982afa30527cbae5de08cb8daR32), but could try removing it to see what breaks because I don't remember :see_no_evil:

File: tweet-api/tests/test_analyze.py (hunk 22 above)
Comment (danasaur) [reply tweet-api/tests/test_analyze.py:12]: Ignore this test directory entirely, I made a task in my initial long comment that I plan to update this. 

Just to rapidly test yesterday, I copied it from another project where "the order of the features for our input vector matter".

File: .gitignore (hunk 4 above)
Comment (danasaur) [reply .gitignore:110]: this is just a "basic python .gitignore template" I copied to remove a bunch of stuff I unintentionally staged. I can clean it up!

File: tweet-api/api/inference.py (hunk 18 above)
Comment (danasaur) [reply tweet-api/api/inference.py:16]: I didn't think of the csv package as "the big guns" and I'm not sure what you're proposing that's simpler. Can you help me understand?

File: tweet-api/api/inference.py (hunk 19 above)
Comment (danasaur) [reply tweet-api/api/inference.py:21]: no, good catch!

**Comment by jpetazzo** [comment]
Also noting that here for further reference: the tokenizer model is downloaded when we `import inference`, but the emotion model is downloaded when we call `predict(...)` the first time.

Both models are cached in `~/.cache/huggingface/transformers`:

```
$ rm -rf ~/.cache/huggingface/transformers/
$ ipython
Python 3.10.8 (main, Nov  1 2022, 14:18:21) [GCC 12.2.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.6.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import inference
Downloading: 100%|██████████████████████████████| 768/768 [00:00<00:00, 446kB/s]
Downloading: 100%|████████████████████████████| 878k/878k [00:02<00:00, 347kB/s]
Downloading: 100%|████████████████████████████| 446k/446k [00:00<00:00, 815kB/s]
Downloading: 100%|█████████████████████████████| 150/150 [00:00<00:00, 93.5kB/s]

In [2]: inference.predict("it is a beautiful day")
Downloading: 100%|███████████████████████████| 476M/476M [00:34<00:00, 14.4MB/s]
Out[2]: 'joy'

In [3]: inference.predict("it is a very beautiful day")
Out[3]: 'joy'

In [4]:                                                                         
Do you really want to exit ([y]/n)? 
$ du -sh ~/.cache/huggingface/transformers/
477M	/home/jp/.cache/huggingface/transformers/
```
